File: sdk/go/ksapp/capability_context.go
package ksapp

import (
	"context"
	"sync/atomic"
	"time"
)

// CapabilityContext 是 capability handler 收到的运行时上下文。
// Handler 签名：func(ctx CapabilityContext, args map[string]any) (any, error)。
type CapabilityContext interface {
	UserID() string
	CallerID() string
	CallerKind() string
	ChainID() string
	ChainHeader() string
	TaskID() string
	RequestID() string
	CanonicalName() string

	// Context 返回原 context.Context，handler 内调下游 SDK / HTTP / 数据库时应透传它，
	// 这样 Task.Cancel / dispatcher 超时 / 上游 ctx done 能直接传到下游链路。
	// 上游未提供时返回 context.Background()（保证非 nil）。
	Context() context.Context

	// Progress 上报进度（仅 long_running 任务有效；sync 调用 / TaskID="" 时 no-op）。
	Progress(ctx context.Context, stage string, percent *int) error
	// Deadline 返回 unix ms；TimeoutMs<=0 返 0。
	Deadline() int64
	// Cancelled cooperative cancellation 检查。
	Cancelled() bool
}

// ProgressReportFn 用于注入实际的 progress 上报路径（DispatcherClient.ReportProgress）。
// 函数注入而非 interface 是为规避循环 import。
type ProgressReportFn func(ctx context.Context, stage string, percent *int) error

// capabilityContextInit 是 newCapabilityContext 的入参集合。
type capabilityContextInit struct {
	Ctx                                                                                  context.Context
	UserID, CallerID, CallerKind, ChainID, ChainHeader, TaskID, RequestID, CanonicalName string
	TimeoutMs                                                                            int
	StartedAtMs                                                                          int64
	ReportFn                                                                             ProgressReportFn
}

// capabilityContextImpl 是 CapabilityContext 的默认实现。
type capabilityContextImpl struct {
	ctx                                                                                  context.Context
	userID, callerID, callerKind, chainID, chainHeader, taskID, requestID, canonicalName string
	timeoutMs                                                                            int
	startedAtMs                                                                          int64
	reportFn                                                                             ProgressReportFn
	cancelled                                                                            atomic.Bool
}

func newCapabilityContext(init capabilityContextInit) *capabilityContextImpl {
	started := init.StartedAtMs
	if started == 0 {
		started = time.Now().UnixMilli()
	}
	ctx := init.Ctx
	if ctx == nil {
		ctx = context.Background()
	}
	return &capabilityContextImpl{
		ctx:           ctx,
		userID:        init.UserID,
		callerID:      init.CallerID,
		callerKind:    init.CallerKind,
		chainID:       init.ChainID,
		chainHeader:   init.ChainHeader,
		taskID:        init.TaskID,
		requestID:     init.RequestID,
		canonicalName: init.CanonicalName,
		timeoutMs:     init.TimeoutMs,
		startedAtMs:   started,
		reportFn:      init.ReportFn,
	}
}

func (c *capabilityContextImpl) UserID() string           { return c.userID }
func (c *capabilityContextImpl) CallerID() string         { return c.callerID }
func (c *capabilityContextImpl) CallerKind() string       { return c.callerKind }
func (c *capabilityContextImpl) ChainID() string          { return c.chainID }
func (c *capabilityContextImpl) ChainHeader() string      { return c.chainHeader }
func (c *capabilityContextImpl) TaskID() string           { return c.taskID }
func (c *capabilityContextImpl) RequestID() string        { return c.requestID }
func (c *capabilityContextImpl) CanonicalName() string    { return c.canonicalName }
func (c *capabilityContextImpl) Context() context.Context { return c.ctx }

func (c *capabilityContextImpl) Progress(ctx context.Context, stage string, percent *int) error {
	if c.taskID == "" || c.reportFn == nil {
		return nil
	}
	// best-effort：业务 handler 不因 progress 失败而失败
	_ = c.reportFn(ctx, stage, percent)
	return nil
}

func (c *capabilityContextImpl) Deadline() int64 {
	if c.timeoutMs <= 0 {
		return 0
	}
	return c.startedAtMs + int64(c.timeoutMs)
}

func (c *capabilityContextImpl) Cancelled() bool {
	return c.cancelled.Load()
}

func (c *capabilityContextImpl) setCancelled() {
	c.cancelled.Store(true)
}

File: sdk/typescript/src/capability_context.ts
/**
 * CapabilityContext：capability handler 收到的运行时上下文。
 * 镜像 Go capability_context.go / Python capability_context.py。
 *
 * TS 适配：Go context.Context（取消/超时/透传）→ TS AbortSignal；cancelled()=signal.aborted；
 * deadline() 返 unix ms。字段 API 面 camelCase；mcp_tool 路径字段源自 _meta.ks_*。
 */

/** progress 上报回调（由 DispatcherClient.reportProgress 注入；避免循环 import）。 */
export type ProgressReporter = (taskId: string, stage: string, percent?: number) => Promise<void>;

export interface CapabilityContext {
  readonly userId: string;
  readonly callerId: string;
  readonly callerKind: string;
  readonly chainId: string;
  readonly chainHeader: string;
  readonly taskId: string;
  readonly requestId: string;
  readonly canonicalName: string;
  /** 透传给下游 fetch / 数据库的取消信号（对齐 Go ctx 透传）。 */
  readonly signal: AbortSignal;
  /** 上报进度（仅 long_running 任务有效；taskId="" 或无 reporter 时 no-op）。 */
  progress(stage: string, percent?: number): Promise<void>;
  /** 任务超时点 unix ms；timeoutMs<=0 返 0。 */
  deadline(): number;
  /** cooperative cancellation 检查（= signal.aborted）。 */
  cancelled(): boolean;
}

export interface CapabilityContextInit {
  userId?: string;
  callerId?: string;
  callerKind?: string;
  chainId?: string;
  chainHeader?: string;
  taskId?: string;
  requestId?: string;
  canonicalName: string;
  timeoutMs?: number;
  startedAtMs?: number;
  reportProgress?: ProgressReporter;
  signal?: AbortSignal;
}

class CapabilityContextImpl implements CapabilityContext {
  readonly signal: AbortSignal;
  private readonly controller?: AbortController;
  private readonly startedAtMs: number;

  constructor(private readonly init: CapabilityContextInit) {
    if (init.signal) {
      this.signal = init.signal;
    } else {
      this.controller = new AbortController();
      this.signal = this.controller.signal;
    }
    this.startedAtMs = init.startedAtMs ?? Date.now();
  }

  get userId(): string { return this.init.userId ?? ""; }
  get callerId(): string { return this.init.callerId ?? ""; }
  get callerKind(): string { return this.init.callerKind ?? ""; }
  get chainId(): string { return this.init.chainId ?? ""; }
  get chainHeader(): string { return this.init.chainHeader ?? ""; }
  get taskId(): string { return this.init.taskId ?? ""; }
  get requestId(): string { return this.init.requestId ?? ""; }
  get canonicalName(): string { return this.init.canonicalName; }

  async progress(stage: string, percent?: number): Promise<void> {
    if (!this.taskId || !this.init.reportProgress) return;
    try {
      await this.init.reportProgress(this.taskId, stage, percent);
    } catch {
      // best-effort：进度丢失不应让业务 handler 失败
    }
  }

  deadline(): number {
    const t = this.init.timeoutMs ?? 0;
    return t <= 0 ? 0 : this.startedAtMs + t;
  }

  cancelled(): boolean {
    return this.signal.aborted;
  }
}

export function createCapabilityContext(init: CapabilityContextInit): CapabilityContext {
  return new CapabilityContextImpl(init);
}

function metaString(meta: Record<string, unknown> | null | undefined, key: string): string {
  if (!meta || typeof meta !== "object") return "";
  const v = (meta as Record<string, unknown>)[key];
  if (v === undefined || v === null) return "";
  return typeof v === "string" ? v : String(v);
}

/**
 * buildContextFromMeta：从 MCP _meta.ks_* 构造 CapabilityContext（mcp_tool backend 生成 tool 路径）。
 * Bug#2：chainHeader 取 ks_chain_snapshot（keystone mcptool executor 透传后非空，见 plan 关键契约 6）。
 */
export function buildContextFromMeta(
  meta: Record<string, unknown> | null | undefined,
  opts: { canonicalName: string; timeoutMs?: number; reportProgress?: ProgressReporter; signal?: AbortSignal },
): CapabilityContext {
  return createCapabilityContext({
    userId: metaString(meta, "ks_user_id"),
    callerId: metaString(meta, "ks_caller_id"),
    callerKind: metaString(meta, "ks_caller_kind"),
    chainId: metaString(meta, "ks_chain_id"),
    chainHeader: metaString(meta, "ks_chain_snapshot"),
    taskId: metaString(meta, "ks_task_id"),
    requestId: metaString(meta, "ks_request_id"),
    canonicalName: opts.canonicalName,
    timeoutMs: opts.timeoutMs,
    reportProgress: opts.reportProgress,
    signal: opts.signal,
  });
}

export interface CallerContext {
  userId: string;
  callerId: string;
  callerKind: string;
  chainId: string;
  chainHeader: string;
  taskId: string;
  requestId: string;
}

/**
 * extractCallerContext：复用降级 helper。复用普通 app.tool 时 handler 拿不到
 * CapabilityContext，caller 上下文在 args._meta.ks_*——本 helper 从中提取。
 */
export function extractCallerContext(args: Record<string, unknown>): CallerContext {
  const meta = (args?._meta ?? {}) as Record<string, unknown>;
  return {
    userId: metaString(meta, "ks_user_id"),
    callerId: metaString(meta, "ks_caller_id"),
    callerKind: metaString(meta, "ks_caller_kind"),
    chainId: metaString(meta, "ks_chain_id"),
    chainHeader: metaString(meta, "ks_chain_snapshot"),
    taskId: metaString(meta, "ks_task_id"),
    requestId: metaString(meta, "ks_request_id"),
  };
}
